File: backend/app/schemas/official.py
"""官网采集接口的请求与响应模型。

**只描述形状，不做业务判断**：能不能采集、结论是什么，都由服务层定好之后填进来。
接口层再算一遍是"前后端阈值漂移"的起点——本项目在投递准入上已经吃过一次，那里靠
``admission_of`` 单源收口，这里同理。

标签字段（``*_label``）与枚举值一起返回，前端直接用中文渲染，不自己维护一份映射表。
"""
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class OfficialSiteCreate(BaseModel):
    company: str = Field(default="", max_length=128)
    homepage_url: str = Field(default="", max_length=512)
    careers_url: str = Field(default="", max_length=512)


class OfficialSiteUpdate(OfficialSiteCreate):
    """修改已保存的公司源；保存后由接口重新探测招聘系统。"""


class CollectRequest(BaseModel):
    """开始一次采集时的可选参数。**整个请求体都是可选的**（不带体 = 按默认来）。

    ``max_jobs`` 是**这一次**的条数上限：岗位上万条的站点一次翻不完，而用户多半只想先看前
    几十条。上限落在服务端、随这次运行走，不存进站点——它回答的是"这回我只要看这么多"。
    """

    # 上限给一个天花板：它就是防手滑的（多打一个零变成抓十万条，界面会显示成"要跑几小时"，
    # 但那时已经点下去了）。真正的规模上限仍是采集器自己那个页数纪律。
    max_jobs: int | None = Field(default=None, ge=1, le=2000)
    # 这次临时筛选的岗位关键词；不写回公司配置。多个关键词可用顿号、逗号、分号或换行分隔。
    job_keywords: str | None = Field(default=None, max_length=200)


class OfficialSiteOut(BaseModel):
    id: int
    company: str
    homepage_url: str
    careers_url: str
    # 识别结果。
    source_kind: str
    # 系统展示名；未识别时为空串（前端据此显示"未识别"而不是显示一个内部标识）。
    source_label: str = ""
    endpoint: str = ""
    confidence: str = ""
    confidence_label: str = ""
    probe_evidence: str = ""
    # robots 结论。``None`` 表示还没查过。
    robots_allowed: bool | None = None
    robots_detail: str = ""
    crawl_delay_seconds: float | None = None
    min_interval_seconds: int = 0
    enabled: bool = True
    last_probed_at: datetime | None = None
    created_at: datetime
    # 列表页要显示的东西，由服务层算好。
    latest_verdict: str = ""
    latest_verdict_label: str = ""
    latest_headline: str = ""
    # 最近一次运行的**状态**。列表页要靠它显示"采集中"——只看 ``latest_verdict`` 是不够的：
    # 采集进行中时结论还是空的，界面会显示成"还没采过"，而用户明明刚点了开始。
    latest_run_id: int | None = None
    latest_status: str = ""
    latest_status_label: str = ""
    # 现在能不能采集：已识别出系统 **且** robots 允许。前端据此禁用按钮，
    # 但服务端在真正执行时仍会自己复核一遍，不信任前端传入的状态。
    can_collect: bool = False


class OfficialRunOut(BaseModel):
    id: int
    site_id: int
    site_company: str = ""
    status: str
    status_label: str = ""
    # ===== 账目 =====
    pages: int = 0
    # 去重后的取回条数；对总数比的是它，不是 stored。
    collected: int = 0
    stored: int = 0
    skipped: int = 0
    detail_missing: int = 0
    total_hint: int | None = None
    # ===== 对账结论 =====
    verdict: str = ""
    verdict_label: str = ""
    evidence: str = ""
    headline: str = ""
    missing: int = 0
    blocks: list[str] = Field(default_factory=list)
    block_labels: list[str] = Field(default_factory=list)
    started_at: datetime
    finished_at: datetime | None = None
    error: str = ""
    # 这次采集花掉的模型调用次数。**用户自付 key，这一项必须可见**——否则他只能去翻服务商的
    # 账单才知道这个功能用了他多少钱。
    llm_calls: int = 0


class OfficialTrendOut(BaseModel):
    """这次采集与这家公司近期基线的偏离。

    **它不是一条对账依据**，而是一个独立提醒：条数下降既可能是我们漏抓，也可能是这家公司真的
    关掉了那些岗位，只看条数是分不出来的。所以 ``detail`` 里要把这句话说清楚。
    """

    state: str
    label: str
    detail: str
    # 基线（历史中位数）与本次条数；历史不足时 baseline 为 null。
    baseline: int | None = None
    latest: int = 0


class OfficialExtractionOut(BaseModel):
    """这次采集**每一页是怎么读出来的**。

    它不是对账依据，而是花销与可复现性的说明：用户能看到哪几页动了模型、配方有没有归纳成功、
    下次还需不需要再花钱。``methods`` 是逐页的句子，由服务层写好，界面原样展示。
    """

    methods: list[str] = Field(default_factory=list)
    llm_calls: int = 0
    # 本次归纳并成功存下的配方数。存下来之后，同一站点的后续采集不再需要模型。
    recipes_learned: int = 0
    # 因为超过层级上限而没有跟进的地址数。**它不改变结论**（站点的次级导航不是岗位），
    # 但报告里要能回答"为什么这次只翻了这么几页"。
    deep_links_skipped: int = 0


class OfficialRunDetailOut(OfficialRunOut):
    """报告页用：带上逐层依据与趋势提醒。

    与列表分开，是因为明细里可能有很多层，列表页不需要它们。
    """

    reconcile_detail: dict[str, Any] = Field(default_factory=dict)
    trend: OfficialTrendOut | None = None
    extraction: OfficialExtractionOut = Field(default_factory=OfficialExtractionOut)


class OfficialProbeOut(BaseModel):
    """一次探测的结果。

    ``state`` 是三态（命中 / 未识别 / 被阻断），**不是布尔**：把"被阻断"混进"未识别"会让
    用户以为这家公司不用这套系统，而实际上我们只是没看到内容。
    """

    state: str
    state_label: str = ""
    site: OfficialSiteOut
    job_count: int = 0
    evidence: str = ""
    # 试过哪些端点、各自什么结果。探测失败时这是用户唯一能拿去反馈的东西。
    attempts: list[dict[str, Any]] = Field(default_factory=list)
    # 注意"没试完"这件事是写在 ``evidence`` 里的（``probe.py`` 会把它挂到结论上），
    # 不再单独开一个字段：**同一句话只该有一处**，而 ``evidence`` 还会落进源的「判断依据」
    # 里长期展示——那里也必须看得到。

File: frontend/src/types/official.ts
/**
 * 官网岗位采集的类型。
 *
 * 枚举值（`verdict` / `status` / `blocks`）与后端的字面量**逐字一致**，中文标签由后端一并
 * 返回（`*_label` 字段）。前端**不自己维护一份映射**——两份映射迟早漂移，而漂移的表现是
 * 某个状态在界面上显示成原始英文，排查时没人会想到去比对两边。
 */

/** 一个公司官网源。 */
export interface OfficialSite {
  id: number;
  company: string;
  homepage_url: string;
  careers_url: string;
  /** 识别出的招聘系统标识；空串 = 未识别。 */
  source_kind: string;
  /** 系统展示名；未识别时为空串。 */
  source_label: string;
  endpoint: string;
  confidence: string;
  confidence_label: string;
  /** 判断依据的说明，直接展示（"为什么认为这家公司用的是这套系统"）。 */
  probe_evidence: string;
  /** robots 结论；null = 还没查过。 */
  robots_allowed: boolean | null;
  robots_detail: string;
  crawl_delay_seconds: number | null;
  min_interval_seconds: number;
  enabled: boolean;
  last_probed_at: string | null;
  created_at: string;
  latest_verdict: string;
  latest_verdict_label: string;
  latest_headline: string;
  /** 最近一次运行的 id；还没采过时为 null。 */
  latest_run_id: number | null;
  /** 最近一次运行的**状态**（running / done / stopped / failed）。 */
  latest_status: string;
  latest_status_label: string;
  /** 现在能不能采集，由服务端判定；前端只读，不二次判断。 */
  can_collect: boolean;
}

/** 一次采集的账目与对账结论。 */
export interface OfficialRun {
  id: number;
  site_id: number;
  site_company: string;
  status: string;
  status_label: string;
  pages: number;
  /** 去重后的取回条数；对总数比的是它。 */
  collected: number;
  stored: number;
  skipped: number;
  detail_missing: number;
  total_hint: number | null;
  verdict: string;
  verdict_label: string;
  evidence: string;
  /** 一句话结论，直接展示。 */
  headline: string;
  /** 已确认的差额条数。 */
  missing: number;
  blocks: string[];
  block_labels: string[];
  started_at: string;
  finished_at: string | null;
  error: string;
  /**
   * 这次采集花掉的模型调用次数。
   *
   * **用户自付 key，这一项必须可见**——否则他只能去翻服务商的账单才知道这个功能用了他多少钱。
   */
  llm_calls: number;
}

/**
 * 这次采集**每一页是怎么读出来的**。
 *
 * 它不是对账依据，而是花销与可复现性的说明：哪几页动了模型、配方有没有归纳成功、下次还需不需要
 * 再花钱。`methods` 是逐页的句子，由后端写好，界面原样展示。
 */
export interface OfficialExtraction {
  methods: string[];
  llm_calls: number;
  /** 本次归纳并成功存下的配方数。存下来之后，同一站点的后续采集不再需要模型。 */
  recipes_learned: number;
  /**
   * 因为超过层级上限而没有跟进的地址数。
   *
   * **它不改变结论**（站点的次级导航不是岗位），但报告里要能回答"为什么这次只翻了这么几页"。
   */
  deep_links_skipped: number;
}

/** 对账报告的一层依据。 */
export interface ReconcileLayer {
  layer: string;
  label: string;
  detail: string;
}

export interface ReconcileDetail {
  basis?: string;
  layers?: ReconcileLayer[];
  termination?: { state: string; detail: string } | null;
  stopped_reason?: string;
  /** 本次临时岗位筛选；空表示没有筛选。 */
  job_filter?: string;
  /** 历史报告里的岗位快照，避免候选后来变化后报告失去可读性。 */
  collected_jobs?: OfficialCollectedJob[];
  robots?: { allowed: boolean; detail: string };
  /**
   * 待核实地址：站点地图里有、这次没抓到的岗位页。
   *
   * **不是"漏抓清单"**——里面既有真漏的，也有早就招满、URL 还留在地图里的历史岗位。
   * 逐个点开看一眼才能分清，所以界面上要把这句话说在前面，不能让用户以为点一遍就对了。
   */
  candidates?: string[];
  /** 逐条核实的结果（点过「核实这些地址」之后才有）。 */
  verifications?: VerificationEntry[];
}

export interface OfficialCollectedJob {
  candidate_id: number | null;
  job_id: number | null;
  title: string;
  company: string;
  location: string;
  source_url: string;
}

/** 一条存活校验的结果。 */
export interface VerificationEntry {
  url: string;
  /** `live` 仍在招 / `gone` 已下架 / `unknown` 判不出来。 */
  state: string;
  /** 中文说明，由后端给出——前端不自己维护枚举映射。 */
  label: string;
  detail: string;
}

/**
 * 这次采集与这家公司近期基线的偏离。
 *
 * **它不是一条对账依据**，而是一个独立提醒：条数下降既可能是我们漏抓，也可能是这家公司真的
 * 关掉了那些岗位——只看条数分不出来，所以 `detail` 里必须把这句话说清楚。
 */
export interface OfficialTrend {
  /** stable / dropped / jumped / insufficient / unmeasured。 */
  state: string;
  /** 中文说明，由后端给出——前端不自己维护枚举映射。 */
  label: string;
  detail: string;
  /** 基线（历史中位数）与本次条数；历史不足时为 null。 */
  baseline: number | null;
  latest: number;
}

/** 报告页用：带上逐层依据、趋势提醒与抽取方式。 */
export interface OfficialRunDetail extends OfficialRun {
  reconcile_detail: ReconcileDetail;
  trend?: OfficialTrend | null;
  extraction: OfficialExtraction;
}

/** 探测时试过的一个端点。 */
export interface OfficialProbeAttempt {
  feed_key: string;
  endpoint: string;
  block: string;
  block_label: string;
  job_count: number;
  detail: string;
}

/**
 * 一次探测的结果。
 *
 * `state` 是三态（命中 / 未识别 / 被阻断），**不是布尔**：把"被阻断"混进"未识别"会让人以为
 * 这家公司不用这套系统，而实际上我们只是没看到内容。
 */
export interface OfficialProbeResult {
  state: string;
  state_label: string;
  site: OfficialSite;
  job_count: number;
  evidence: string;
  attempts: OfficialProbeAttempt[];
}

export interface OfficialSitePayload {
  company: string;
  homepage_url?: string;
  careers_url?: string;
}

/**
/** 采集结论 → 展示用的语义色。**只影响颜色，不影响措辞**（措辞全部来自后端）。 */
export const VERDICT_COLORS: Record<string, string> = {
  complete: "success",
  incomplete: "warning",
  unknown: "default",
};
